Feat named imports

tofarr opened the pull request:
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [X] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
This PR introduces the concept of importing implementations by name - something the SAAS offering will take advantage of. Effectively, the pattern will be to allow overriding existing behavior at set pivot points to match the environment's requirements.


---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:69d3992-nikolaik   --name openhands-app-69d3992   docker.all-hands.dev/all-hands-ai/openhands:69d3992
```

enyst commented on openhands/server/app.py:
```diff
@@ -48,6 +49,7 @@ async def health():
 app.include_router(security_api_router)
 app.include_router(feedback_api_router)
 
+AttachSessionMiddleware = import_from(config.attach_session_middleware_class)
```
There's a lot to like about dynamically loading stuff (I'm guilty of a love for metaclasses even), but this bit is a little confusing to me. I'm concerned that it makes the code less clear, it's less understandable at a glance what is running and how come. The other PR (on refactoring github auth) adds this kind of imports to more places, too. The code seems to get quite a number of 'if this then load the other'.

Could we implement this via an interface or default class, where github auth is plugged in a single place perhaps?

tofarr replied:
Agreed - a base class makes sense for this.

tofarr commented:
So given that folks seemed to be concerned about the lack of type safety in the `import_from` function, I introduced a `get_impl` function which ensures that the returned class is the same as or a subclass of a specified value. I also added a quick test to validate the import.

The type checker in PyCharm is smart enough to figure this out an point back to the specified base class when you cmd+click, but the type checker in vs-code is not.

Engel did voice concern that this makes the connections a little bit more indirect, but this is by design - there will be cases where functionality is required for the SAAS env and it does not make sense to pollute the Open Source code base with things that are not relevant to it. I think the `get_impl` approach strikes a reasonable balance between these concerns, and leaves the open source codebase as clean as possible.

rbren reviewed (approved):
This looks great! Love that we kept some type safety
